### PR TITLE
feat: Microsoft.Bot.Core Runtime Configuration

### LINKED ISSUE #2: Definition of Core library
.NET code project for a bot available as a library, this does not produce an application on itself.

- [x] Rename namespace/classes to Core/Runtime appropriate
- [x] Pull out and generalize Composer-specific settings
- [ ] Enables low-code startup of host application using standardized application configuration

### PR DESCRIPTION
**Pull request by peterinnesmsft**
Fixes #2, #4, #14

## Description
This change implements a number of major changes to Microsoft.Bot.Core to handle configuration of the runtime host application via a new runtime.json file, which can bind settings to values defined through standard application settings provided through supported methods (appsettings.json, environment variables, etc.).

## Specific Changes

  - Removes a number of settings classes that were not being utilized, while introducing new options classes that are leveraged by updates made to both CoreBot and CoreBotAdapter.
  - Defining IProvider and IBuilder interfaces and a variety of implementations around core SDK components to handle the registration of various services, as well as provide instances of objects such as middleware.
  - Provided new extension classes for IServiceCollection and ExpressionProperty<T>, as well as centralized all IConfigurationBuilder extensions into a single class.

## Testing
Unit tests have been written to exercise all major new components added to Microsoft.Bot.Core. The focus of these is primarily to verify registered service types within the host application's service container, as well as ensure that they can be successfully instantiated, as based upon provided configuration.

### CODE REVIEW

**Review by ryanlengel — CHANGES_REQUESTED** [review]

File: src/Microsoft.Bot.Core/AdaptiveConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+// Licensed under the MIT License.
+
+using System;
+using System.Globalization;
+using System.Text.RegularExpressions;
+using Microsoft.Extensions.Configuration;
+
+namespace Microsoft.Bot.Core
+{
+    public class AdaptiveConfiguration
```
[the remark this answers is not in the document]
Comment (peterinnesmsft) [reply src/Microsoft.Bot.Core/AdaptiveConfiguration.cs:11]: The purpose of this class is to provide a wrapper that conforms to the IMemory interface defined in AdaptiveExpressions.Memory.IMemory, which allows it to be bound to a given adaptive expression - in other words, this is the thing that allows us to specify adaptive expressions that bind to IConfiguration.

To align, I could rename to ConfigurationMemory and put in suitable documentation tags?

File: src/Microsoft.Bot.Core/Builders/Transcripts/BlobsTranscriptStoreBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+using AdaptiveExpressions.Properties;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.Azure;
+using Microsoft.Bot.Core.Extensions;
+using Microsoft.Extensions.Configuration;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Core.Builders.Transcripts
+{
+    // TODO: Change parent interface to ITranscriptStoreBuilder
```
[the remark this answers is not in the document]
Comment (peterinnesmsft) [reply src/Microsoft.Bot.Core/Builders/Transcripts/BlobsTranscriptStoreBuilder.cs:14]: Open #39 to track.

File: src/Microsoft.Bot.Core/Builders/Transcripts/BlobsTranscriptStoreBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            // TODO: Change to Microsoft.Bot.Builder.Azure.BlobsTranscriptStore
```
[the remark this answers is not in the document]
Comment (peterinnesmsft) [reply src/Microsoft.Bot.Core/Builders/Transcripts/BlobsTranscriptStoreBuilder.cs:39]: This is a slightly different issue, so I've created #40 as a separate tracking issue.

File: src/Microsoft.Bot.Core/Providers/Storage/BlobStorageProvider.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            // TODO: Change to Microsoft.Bot.Builder.Azure.BlobsStorage
+            services.AddSingleton<IStorage>(new AzureBlobStorage(
```
[the remark this answers is not in the document]
Comment (peterinnesmsft) [reply src/Microsoft.Bot.Core/Providers/Storage/BlobStorageProvider.cs:40]: Created #40 to track.

**Pull request merged**